tests/Test/AbstractIntegrationTestCase.php

```diff
@@ -146,6 +146,20 @@ public function testIntegration(IntegrationCase $case): void
         }
 
         $this->doTest($case);
+
+        // run the test again with the `expected` part, this should always stay the same
+        $this->doTest(
```

Comment:
i'm mid-mid here.

this approach means whenever someone calling `->doTest()`, they shall always call it twice. I'm not sure if it's good.
(i know there are not so many cases we call it, but imagine we make this AbstractIntegrationTestCase public ultimately, for libraries of custom repos)


maybe they can continue to call it once, and in doTest we do like
```
function doTest() {
    check reqs for PHP...

    $this->verifyIntegration($expected, $input); // of course, full IntegrationCase to be passed
    $this->verifyIntegration($expected, null);
}

Reply:
Wasn't the previous implementation even worse? Anyone calling `doTest` would also call `testIntegration`. We can refactor this later, this one is only about removing recursion.

Reply:
> Wasn't the previous implementation even worse?

thus I gave the approval, right?